File: hw2/src/container/Container.cpp
#include "Container.h"

Container::~Container() noexcept {
    size_ = 0;
    data.clear();
    std::cout << "Container cleared and deleted";
}

void Container::setSize(int size) {
    size_ = size;
}

void Container::readData(std::ifstream &input) {
    for (int i = 0; i < size_; i++) {
        int type, first_param, sec_param, third_param;
        double density;
        input >> type >> density;

        // Determine which figure read.
        switch (type) {

            // 1 - sphere.
            case 1:
                input >> first_param;
                data.push_back(std::make_unique<Sphere>(density, first_param));
                break;

            // 2 - parallelepiped.
            case 2:
                input >> first_param >> sec_param >> third_param;
                data.push_back(std::make_unique<Parallelepiped>(density, first_param, sec_param, third_param));
                break;

            // 3 - tetrahedron.
            case 3:
                input >> first_param;
                data.push_back(std::make_unique<Tetrahedron>(density, first_param));
                break;

            // If is the type incorrect.
            default:
                std::cout << "Incorrect type of shape. Bye!";
                exit(-1);
        }
    }
}

void Container::rndData() {
    for (int i = 0; i < size_; i++) {
        int type = randomInt() % 3 + 1;
        int first_param, sec_param, third_param;
        // Set a random value.
        double density = randomDouble();

        // Determine which figure read.
        switch (type) {

            // 1 - sphere.
            case 1:
                first_param = randomInt();
                data.push_back(std::make_unique<Sphere>(density, first_param));
                break;

            // 2 - parallelepiped.
            case 2:
                first_param = randomInt();
                sec_param = randomInt();
                third_param = randomInt();
                data.push_back(std::make_unique<Parallelepiped>(density, first_param, sec_param, third_param));
                break;

            // 3 - tetrahedron.
            case 3:
                first_param = randomInt();
                data.push_back(std::make_unique<Tetrahedron>(density, first_param));
                break;

            // If is the type incorrect.
            default:
                std::cout << "Incorrect type of shape. Bye!";
                exit(-1);
        }
    }
}

void Container::writeData(std::ofstream &output) {
    output << "There are " << size_ << " shapes in the container.\n";
    output << "\n";
    for (int i = 0; i < size_; i++) {
        output << i + 1 << "  ----------  ";
        // Write a message about the shape.
        data[i]->writeData(output);
        output << "\n";
    }
}

void Container::sortData() {
    int size = size_;
    int start = 0, end = 0;
    for (int i = 1; i < size; ++i) {

        // Select the current element.
        std::unique_ptr<Shape> key = std::move(data[i]);

        start = 0;
        end = i - 1;

        // Using the method of dividing in half,
        // we select an element whose density is less.
        while (start <= end) {

            int mid = start + (end - start) / 2;

            if (key->density() > data[mid]->density()) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        // Swap all elements we determine.
        for (int j = i - 1; j >= start; --j) {
            swap(data[j + 1], data[j]);
        }

        // Insert the current element.
        data[start] = std::move(key);
    }
}

File: hw3/src/container/container.py
from src.shapes.paralellepiped import Parallelepiped
from src.shapes.sphere import Sphere
from src.shapes.tetrahedron import Tetrahedron
from src.random.random import random_int, random_double


class Container:

    def __init__(self):
        self.__shapes__ = []
        self.__size__ = 0
        pass

    def read_data(self, input_stream):
        for i in range(self.__size__):
            first_param = -1
            second_param = -1
            third_param = -1
            density = -1
            my_shape_params = input_stream.readline().split()
            type_of_shape = int(my_shape_params[0])

            if type_of_shape == 1:

                if len(my_shape_params) != 3:
                    print("Incorrect number of params... Bye!")
                    exit(-1)

                density = float(my_shape_params[1])
                first_param = int(my_shape_params[2])
                self.__shapes__.append(Sphere(density, first_param))

            elif type_of_shape == 2:

                if len(my_shape_params) != 5:
                    print("Incorrect number of params... Bye!")
                    exit(-1)

                density = float(my_shape_params[1])
                second_param = int(my_shape_params[2])
                first_param = int(my_shape_params[3])
                third_param = int(my_shape_params[4])
                self.__shapes__.append(Parallelepiped(density, first_param, second_param, third_param))

            elif type_of_shape == 3:

                if len(my_shape_params) != 3:
                    print("Incorrect number of params... Bye!")
                    exit(-1)

                density = float(my_shape_params[1])
                first_param = int(my_shape_params[2])
                self.__shapes__.append(Tetrahedron(density, first_param))
            else:
                print("Incorrect type of shape.")
        pass

    def rnd_data(self):
        for i in range(self.__size__):
            type_of_shape = random_int() % 3 + 1

            if type_of_shape == 1:
                density = random_double()
                first_param = random_int()
                self.__shapes__.append(Sphere(density, first_param))

            elif type_of_shape == 2:
                density = random_double()
                second_param = random_int()
                first_param = random_int()
                third_param = random_int()
                self.__shapes__.append(Parallelepiped(density, first_param, second_param, third_param))

            elif type_of_shape == 3:
                density = random_double()
                first_param = random_int()
                self.__shapes__.append(Tetrahedron(density, first_param))

            else:
                print("Incorrect type of shape.")
        pass

    def sort_data(self):
        for i in range(1, self.__size__):
            key = self.__shapes__[i]
            start = 0
            end = i - 1

            start = self.find_index(key, start, end)

            for j in range(i - 1, start, -1):
                temp = self.__shapes__[j + 1]
                self.__shapes__[j + 1] = self.__shapes__[j]
                self.__shapes__[j] = temp

            self.__shapes__[start] = key
        pass

    def set_size(self, size):
        if size <= 0:
            print("Incorrect size. Bye!\n")
            exit(-1)
        self.__size__ = size
        pass

    def write_data(self, output_stream):
        output_stream.write("There are " + str(self.__size__) + " shapes in the container.\n\n")
        for i in range(self.__size__):
            output_stream.write(str(i + 1) + "  ----------  ")
            output_stream.write(self.__shapes__[i].write_data() + "\n")

    def find_index(self, key, start, end):
        if end <= start:
            if key.density() < self.__shapes__[start].density():
                return start + 1
            return start

        mid = (start + end) // 2

        if key.density() < self.__shapes__[mid].density():
            return self.find_index(key, mid + 1, end)

        if key.density() == self.__shapes__[mid].density():
            return mid + 1

        return self.find_index(key, start, mid - 1)
